### PR TITLE
policy: remove checking of CIDR-based fields from `IsLabelBased` checks

### PR DESCRIPTION
**Pull request by ianvernon**
CIDR-based policy is converted to a label-based representation, so CIDR-policy
is in a sense label-based. Update the `IsLabelBased` functions for ingress and
egress rules, as well as add unit tests accordingly. The removal of checking of
CIDR-based fields in the `IsLabelBased` checks means that we now do wildcarding
at L7 for CIDR-based rules as well in addition to `EndpointSelector` based
rules. Policies like the following will now result in wildcarding at L7:

```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
description: "L3-L4-L7 CIDR policy"
metadata:
  name: "l7-cidr-cnp"
spec:
  egress:
  - toCIDR:
    - 1.2.3.4/32
    toPorts:
    - ports:
      - port: "80"
        protocol: TCP
      rules:
        http:
        - headers:
          - 'X-My-Header: true'
          method: GET
          path: /
  endpointSelector:
    matchLabels:
      id: app1
```

```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
description: "L3 CIDR policy"
metadata:
  name: "l3-cidr-cnp"
spec:
  egress:
  - toCIDR:
    - 1.2.3.4/32
  endpointSelector:
    matchLabels:
      id: app1
```

In other words, even though there is one rule (l7-cidr-cnp) which only allows
egress access to `GET /` with `X-My-Header:true` to `1.2.3.4/32` on port 80/TCP,
the second rule (l3-cidr-cnp) pokes a wider hole in the firewall to allow all
egress traffic to 1.2.3.4. However, we still will redirect to the proxy, but
we will allow all traffic to preserve existing behavior for visibility purposes.

Note that rules with `ToServices` and `ToFQDNs` are not considered label based
as of this commit.

Signed-off by: Ian Vernon <ian@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9318)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Build failure: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/15228/

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]
LGTM to fix the current issue here.

I still don't follow what it means for the "IsLabelBased()" check and how that's supposed to influence policy generation, in the end all policy ends up being label-based. The new unit tests for this test the behaviour you've changed it to, which I guess is better than no tests, but it's not immediately obvious from a high level that those are the correct long-term values for these functions.

I'm glad to see some additional wildcarding tests at the end here, looks right.

**Comment by ianvernon** [comment]
test-me-please

**Comment by tgraf** [comment]
It looks like we should be removing this function entirely? Even the remaining rule types for whih `IsLabelsBased` still returns false should be wildcarded?

**Comment by tgraf** [comment]
Discussed offline. `IsLabelBased` should likely be renamed for clarity. This PR is OK while keeping the changes minimal.

**Review by tgraf — APPROVED** [review]
